text/0000-externally-definable-statics.md

```diff
@@ -0,0 +1,186 @@
+
+The compiler of the root crate will find the implementation of all externally definable statics and give an error
+if more than one definition is found for any of them.
+If none are found, the result is either an error, or—if the `extern static` has a default value—a definition is emitted with that value.
+
+# Drawbacks
+
+- It encourages globally defined behaviour.
+  - Counterargument: We are already doing this anyway, both inside the standard library (e.g. panic_handler, allocator)
+    and outside (e.g. global logger). This just makes it much easier (and safer) to get right.
+- This will invite the addition of many hooks to the standard library to modify existing behavior.
+  While we should consider such possibilities, this RFC does not propose that every piece of standard library behavior should be replaceable.
+
+# Rationale and alternatives
+
+## Syntax
+
+The syntax re-uses existing keywords. Alternatively, we could:
+  - Use the `override` reserved keyword.
+  - Add a new (contextual) keyword (e.g. `existential static`, `define static`, …).
+  - Use an attribute (e.g. `#[externally_definable]`) instead.
+
+## Functions or statics
+
+This RFC only proposes externally definable *statics*.
+
+An alternative is to only provide externally implementable *functions* instead ([RFC 3632](https://github.com/rust-lang/rfcs/pull/3632)).
+
+That would be equivalent in power: one can store a function pointer in a static, and one can return a reference to a static from a function.
+
+(Another alternative, of course is to provide both. See future possibilities.)
+
+## Visibility
+
+There are two kinds of visibilities to be considered for externally definable statics:
+who can *define* the static, and who can *access* the static.
+
+Not allowing the static to be defined by other crates nullifies the functionality, as the entire point of externally definable statics is that they can be defined in another crate. This visibility is therefore always (implicitly) "pub".
+
+Allowing a more restricted (that is, not `pub`) visibility for *accessing* the static can be useful.
+For example, today's `#[panic_handler]` can be defined by any crate, but can not be called directly. (Only indirectly through `panic!()` and friends.)
+
+A downside is that it is not possible to allow this "only definable but not publicly accessible" visibility through an alias.
+
+An alternative could be to use the same visibility for both defining an accessing, which would simply mean that the static (or an alias to it) will always have to be `pub`.
+
+# Prior art
+
+[RFC 2494 "Existential types with external definition"](https://github.com/rust-lang/rfcs/pull/2492)
+has been proposed before, which basically does this for *types*. Doing this for statics (as a start) saves a lot of complexity.
+
+# Unresolved questions
+
+- Should we allow some form of subtyping, similarly to how traits allow trait impls to do subtyping?
```

Comment:
I think we should, but I think it'd be fine to move this to "future work" rather than requiring this in the initial version. It's a backwards-compatible addition, and *not* allowing subtyping for now doesn't close the door for doing it in the future.